packages/hardhat-viem/src/internal/tasks.ts

```diff
@@ -0,0 +1,271 @@
+  getFullyQualifiedName,
+  parseFullyQualifiedName,
+} from "hardhat/utils/contract-names";
+import { getAllFilesMatching } from "hardhat/internal/util/fs-utils";
+
+interface EmittedArtifacts {
+  artifactsEmittedPerFile: ArtifactsEmittedPerFile;
+}
+
+/**
+ * This override generates an artifacts.d.ts file that's used
+ * to type hre.artifacts.
+ *
+ * TODO: Can we avoid regenerating this every time? The reason we override
+ * this task is that deleting a `.sol` file doesn't emit any artifact, yet
+ * we may need to regenerate this file.
+ */
+subtask(TASK_COMPILE_SOLIDITY).setAction(
+  async (_, { config, artifacts }, runSuper) => {
+    const superRes = await runSuper();
+
+    const fqns = await artifacts.getAllFullyQualifiedNames();
+    const contractNames = fqns.map(
+      (fqn) => parseFullyQualifiedName(fqn).contractName
+    );
+
+    const artifactsDTs = generateArtifactsDefinition(contractNames);
+
+    try {
+      await writeFile(
+        join(config.paths.artifacts, "artifacts.d.ts"),
+        artifactsDTs
+      );
+    } catch (error) {
+      console.error("Error writing artifacts definition:", error);
+    }
+
+    return superRes;
+  }
+);
+
+/**
+ * This override generates a .ts file per contract, and a file.d.ts
+ * per solidity file, which is used in conjunction to artifacts.d.ts
+ * to type hre.artifacts.
+ */
+subtask(TASK_COMPILE_SOLIDITY_EMIT_ARTIFACTS).setAction(
+  async (_, { artifacts, config }, runSuper): Promise<EmittedArtifacts> => {
+    const { artifactsEmittedPerFile }: EmittedArtifacts = await runSuper();
+
+    const fqns = await artifacts.getAllFullyQualifiedNames();
+    const contractNames = fqns.map(
+      (fqn) => parseFullyQualifiedName(fqn).contractName
+    );
+    const dupContractNames = contractNames.filter(
+      (name, i) => contractNames.indexOf(name) !== i
+    );
+
+    await Promise.all(
+      artifactsEmittedPerFile.map(async ({ file, artifactsEmitted }) => {
+        const srcDir = join(config.paths.artifacts, file.sourceName);
+        await mkdir(srcDir, {
+          recursive: true,
+        });
+
+        const contractTypeData = await Promise.all(
+          artifactsEmitted.map(async (contractName) => {
+            const fqn = getFullyQualifiedName(file.sourceName, contractName);
+            const artifact = await artifacts.readArtifact(fqn);
+            const isDup = dupContractNames.includes(contractName);
+            const declaration = generateContractDeclaration(artifact, isDup);
+
+            const typeName = `${contractName}$Type`;
+
+            return { contractName, fqn, typeName, declaration };
+          })
+        );
+
+        const fp: Array<Promise<void>> = [];
+        for (const { contractName, declaration } of contractTypeData) {
+          fp.push(writeFile(join(srcDir, `${contractName}.ts`), declaration));
+        }
+
+        const indexTs = generateIndexFile(contractTypeData);
+        fp.push(writeFile(join(srcDir, "index.ts"), indexTs));
```

Comment:
I think that was because we had ts code in some of them. But now it's all type declarations, so it makes sense to change the extension to d.ts. Additionally, I've renamed some of the file names for clarity.